### ISSUE TITLE
`encoding.only_fields` fails to remove array fields

### ISSUE THREAD

**Issue by jszwedko** [vectordotdev/vector#7322]
### Vector Version

```
vector 0.13.0 (v0.13.0 x86_64-unknown-linux-gnu 2021-04-21)
```

### Vector Configuration File

```toml
[sources.in]
type = "stdin"

[transforms.remap]
type = "remap"
inputs = ["in"]
source = """
. |= object!(parse_json!(.message))
"""

[sinks.out]
type = "console"
inputs = ["remap"]
encoding.codec = "json"
encoding.only_fields = ["foo"]
```

### Debug Output

```
➜  vector git:(emit-multiple) ✗ echo '{ "foo": 2, "bar": [1,2,3], "baz": true}' | vector-0.13.0 --config /tmp/3.toml
May 04 16:23:58.791  INFO vector::app: Log level is enabled. level="vector=info,codec=info,vrl=info,file_source=info,tower_limit=trace,rdkafka=info"
May 04 16:23:58.791  INFO vector::sources::host_metrics: PROCFS_ROOT is unset. Using default '/proc' for procfs root.
May 04 16:23:58.791  INFO vector::sources::host_metrics: SYSFS_ROOT is unset. Using default '/sys' for sysfs root.
May 04 16:23:58.791  INFO vector::app: Loading configs. path=[("/tmp/3.toml", None)]
May 04 16:23:58.795  INFO vector::sources::stdin: Capturing STDIN.
May 04 16:23:58.795  INFO vector::topology: Running healthchecks.
May 04 16:23:58.795  INFO vector::topology: Starting source. name="in"
May 04 16:23:58.795  INFO vector::topology: Starting transform. name="remap"
May 04 16:23:58.795  INFO vector::topology: Starting sink. name="out"
May 04 16:23:58.795  INFO vector::topology::builder: Healthcheck: Passed.
May 04 16:23:58.795  INFO vector: Vector has started. version="0.13.0" git_version="v0.13.0" released="Wed, 21 Apr 2021 20:34:28 +0000" arch="x86_64"
May 04 16:23:58.795  INFO vector::app: API is disabled, enable by setting `api.enabled` to `true` and use commands like `vector top`.
May 04 16:23:58.795  INFO vector::shutdown: All sources have finished.
May 04 16:23:58.795  INFO source{component_kind="source" component_name=in component_type=stdin}: vector::sources::stdin: Finished sending.
May 04 16:23:58.795  INFO vector: Vector has stopped.
May 04 16:23:58.796  INFO vector::topology: Shutting down... Waiting on running components. remaining_components="remap, out" time_remaining="59 seconds left"
{"bar":[null,null],"foo":2}
```

### Expected Behavior

`bar` would be removed

### Actual Behavior

`bar`'s subkeys are set to `null`

**Comment by jeromekleinen-kbc** [comment]
Please note that this also affects objects. Consider the following input:
```
{"a": {"b":"1"}, "foo":"bar"}
```

Results in:
```
{"a":{},"foo":"bar"}
```

vector 0.13.1 x86-64

**Comment by jszwedko** [comment]
User in discord reported they were still seeing this in 0.15.1: https://discord.com/channels/742820443487993987/746070591097798688/874649140146089985

**Comment by jszwedko** [comment]
Closed by #8645
